src/core/text-buffer.cc

```diff
@@ -255,57 +256,55 @@ struct TextBuffer::Layer {
-              }
-              break;
+        MatchResult match_result = regex.match(slice_to_search.data(), slice_to_search.size());
+        switch (match_result.type) {
+          case MatchResult::Error:
+            chunk_continuation.clear();
+            return true;
+
+          case MatchResult::None:
+            slice_to_search_start_offset += slice_to_search.size();
+            slice_to_search_start_position = slice_to_search_start_position.traverse(
+              slice_to_search.extent()
+            );
+            chunk_continuation.clear();
+            break;
+
+          case MatchResult::Partial:
+            if (chunk_continuation.empty() || match_result.start_offset > 0) {
+              Point partial_match_position = slice_to_search.position_for_offset(match_result.start_offset);
+              slice_to_search_start_offset += match_result.start_offset;
+              slice_to_search_start_position = slice_to_search_start_position.traverse(partial_match_position);
+              chunk_continuation.assign(slice_to_search.suffix(partial_match_position));
             }
-
-            default:
-              return true;
-          }
-        } else {
-          uint32_t start_offset = regex.get_match_offset(0);
-          uint32_t end_offset = regex.get_match_offset(1);
-
-          result = Range{
-            slice_to_search_start_position.traverse(
-              slice_to_search.position_for_offset(start_offset)
-            ),
-            slice_to_search_start_position.traverse(
-              slice_to_search.position_for_offset(end_offset)
-            )
-          };
-
-          // If the match ends with a CR at the end of a chunk, continue looking
-          // at the next chunk, in case that chunk starts with an LF. Points
-          // within CRLF line endings are not valid.
-          if (end_offset == slice_to_search.size() && slice_to_search.back() == '\r') continue;
-
-          return true;
+            break;
+
+          case MatchResult::Full:
+            result = Range{
+              slice_to_search_start_position.traverse(
+                slice_to_search.position_for_offset(match_result.start_offset)
+              ),
+              slice_to_search_start_position.traverse(
+                slice_to_search.position_for_offset(match_result.end_offset)
+              )
+            };
+
+            // If the match ends with a CR at the end of a chunk, continue looking
+            // at the next chunk, in case that chunk starts with an LF. Points
+            // within CRLF line endings are not valid.
+            if (match_result.end_offset == slice_to_search.size() && slice_to_search.back() == '\r') continue;
+
+            return true;
         }
       }
 
       return false;
     });
 
+    if (!result && !chunk_continuation.empty()) {
+      result = Range{slice_to_search_start_position, extent()};
+    }
+
```

Comment:
@nathansobo I wasn't handling partial matches that occurred at the EOF. When an expression like `a*` matches a suffix of the buffer, pcre will return `PCRE2_ERROR_PARTIAL`, because it doesn't know that it's at the final chunk of input.

So we just need to check if we had a partial match after completing our `for_each_chunk_in_range` loop.